pkg/reconciler/reconciler_test.go

```diff
@@ -1545,6 +1508,181 @@ var _ = Describe("Reconciler", func() {
+		var (
+			ctx                   context.Context
+			cancel                context.CancelFunc
+			matchingLabels        map[string]string
+			anotherMatchingLabels map[string]string
+			mu                    sync.Mutex
+			doneChans             []chan struct{}
+		)
+
+		BeforeEach(func() {
+			matchingLabels = map[string]string{"app": "foo"}
+			anotherMatchingLabels = map[string]string{"app": "bar"}
+			doneChans = nil
+			ctx, cancel = context.WithCancel(context.Background())
+
+			cleanupMgr := getManagerOrFail()
+			cleanupClient := cleanupMgr.GetClient()
+			crList := &unstructured.UnstructuredList{}
+			crList.SetGroupVersionKind(gvk)
+			Expect(cleanupClient.List(ctx, crList)).To(Succeed())
+
+			for i := range crList.Items {
+				cr := &crList.Items[i]
+				// Remove all finalizers to allow CR deletion
+				cr.SetFinalizers([]string{})
+				Expect(cleanupClient.Update(ctx, cr)).To(Succeed())
+				Expect(cleanupClient.Delete(ctx, cr)).To(Succeed())
+			}
+
+			// Wait for all CRs to be deleted
+			Eventually(func() bool {
+				crList := &unstructured.UnstructuredList{}
+				crList.SetGroupVersionKind(gvk)
+				if err := cleanupClient.List(ctx, crList); err != nil {
+					return false
+				}
+				return len(crList.Items) == 0
+			}, "10s", "100ms").Should(BeTrue())
+		})
+
+		AfterEach(func() {
+			// Cancel the context to stop all managers
+			if cancel != nil {
+				cancel()
+			}
+			// Wait for all managers to shut down completely
+			for _, done := range doneChans {
+				select {
+				case <-done:
+					// Manager has shut down
+				case <-time.After(5 * time.Second):
+					// Timeout waiting for manager shutdown
+				}
+			}
+		})
+
+		It("should only reconcile CRs matching the label selector", func() {
+			labeledObj := testutil.BuildTestCR(gvk)
+			labeledObj.SetName("labeled-cr-test1")
+			labeledObj.SetLabels(matchingLabels)
+			labeledObjKey := types.NamespacedName{Namespace: labeledObj.GetNamespace(), Name: labeledObj.GetName()}
+
+			anotherObj := testutil.BuildTestCR(gvk)
+			anotherObj.SetName("another-cr-test1")
+			anotherObjKey := types.NamespacedName{Namespace: anotherObj.GetNamespace(), Name: anotherObj.GetName()}
+
+			var reconciledCRs []string
+			postHook := makePostHook(&mu, &reconciledCRs)
+			mgr, done := setupManagerWithSelectorAndPostHook(ctx, postHook, matchingLabels)
+			doneChans = append(doneChans, done)
+
+			By("creating a CR without matching labels", func() {
+				Expect(mgr.GetClient().Create(ctx, anotherObj)).To(Succeed())
+			})
+
+			By("verifying that the labeled reconciler does not reconcile CR without labels", func() {
+				Consistently(func() bool {
+					mu.Lock()
+					defer mu.Unlock()
+					return len(reconciledCRs) == 0
+				}, "2s", "100ms").Should(BeTrue())
+			})
+
+			By("creating a CR with matching labels", func() {
+				Expect(mgr.GetClient().Create(ctx, labeledObj)).To(Succeed())
+			})
+
+			By("verifying only the labeled CR was reconciled", func() {
+				Eventually(func() bool {
```

Comment:
is it possible that the non-labeled object gets reconciled straight after and we don't catch it?

Reply:
Yes, unfortunately [there does not seem to be a way](https://kubernetes.slack.com/archives/C02MRBMN00Z/p1762251589740719) to know whether controller-runtime will never do something, or just didn't get round to doing it yet :-(

Reply:
It might be worth calling this out in a comment somewhere. In case we hit a flake, we can get the context.

Reply:
I think it would be fine to just add a comment to that effect and we move forward with the PR. I don't think it would be be highly likely to happen, and I'm struggling to think of another test strategy to avoid this.

Reply:
Added a comment